### PR TITLE
Origin/users/amitjoshi/content snippet

### PR DESCRIPTION
**Pull request by amitjoshi438**
This PR integrates content snippet creation in VSCode using Yeoman Generator:
contensnippetCreation

### CODE REVIEW

File: src/client/powerpages/Contentsnippet.ts [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        terminal.sendText(`cd data\n ../node_modules/.bin/yo @microsoft/powerpages:contentsnippet "${name}" "${type}"`);
+
+        const folder = folderName(name);
+        const file = fileName(name);
+
+        const watcher = vscode.workspace.createFileSystemWatcher(
+            new vscode.RelativePattern(
+                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                vscode.workspace.workspaceFolders![0],
+                `**/content-snippets/${folder}/${file}.en-US.contentsnippet.yml`
```
Comment (ashishchoudhary001) [inline src/client/powerpages/Contentsnippet.ts:31]: the language value here en-US shouldn't be hardcoded.

File: src/client/powerpages/Contentsnippet.ts [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+import * as vscode from "vscode";
+import { fileName, folderName, isNullOrEmpty } from "./utils/CommonUtils";
+import { QuickPickItem } from "vscode";
+import { MultiStepInput } from "./utils/MultiStepInput";
+
+export const contentSnippet = async (context: vscode.ExtensionContext) => {
+    const { name, type } = await myMultiStepInput();
+
+    if (!isNullOrEmpty(name)) {
+        const terminal = vscode.window.createTerminal("Powerpages", "");
```
Comment (BidishaMS) [inline src/client/powerpages/Contentsnippet.ts:21]: Should'nt this be localized?

File: src/client/powerpages/Contentsnippet.ts [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+            watcher.onDidCreate(async (uri) => {
+                await vscode.window.showTextDocument(uri);
+                terminal.dispose();
+            });
+        });
+    }
+};
+
+
+async function myMultiStepInput() {
```
Comment (BidishaMS) [inline src/client/powerpages/Contentsnippet.ts:54]: Please rename the method

File: src/client/powerpages/Contentsnippet.ts [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+        (label) => ({ label })
+    );
+
+    interface State {
+        title: string;
+        step: number;
+        totalSteps: number;
+        type: QuickPickItem | string;
+        name: string;
+    }
```
Comment (BidishaMS) [inline src/client/powerpages/Contentsnippet.ts:65]: Looks like generic interface. Please make a separate file for this

File: src/client/powerpages/utils/MultiStepInput.ts [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+                this.current = input;
+                this.current.show();
+            });
+        } finally {
+            disposables.forEach((d) => d.dispose());
+        }
+    }
+
+    async showInputBox<P extends InputBoxParameters>({
+        title,
```
Comment (BidishaMS) [inline src/client/powerpages/utils/MultiStepInput.ts:126]: Naming convention looks very naive in most of the method

File: src/client/powerpages/Contentsnippet.ts [hunk 6]
```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (BidishaMS) [inline src/client/powerpages/Contentsnippet.ts:2]: ContentSnippet does not looks descriptive. How we name it as ContentSnippetOperations

File: src/client/powerpages/utils/CommonUtils.ts [hunk 7]
```diff
@@ -0,0 +1,19 @@
+/*
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See License.txt in the project root for license information.
+ */
+
+export const isNullOrEmpty = (str: string | undefined): boolean => {
```
Comment (BidishaMS) [inline src/client/powerpages/utils/CommonUtils.ts:6]: There is a significant difference in null and undefined in JS. 
Keep a method for the check null and undefined.
keep a method for empty length string.

Incase you are interested to understand the difference https://www.geeksforgeeks.org/undefined-vs-null-in-javascript/

File: src/client/powerpages/utils/CommonUtils.ts [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    return !str || str.trim().length === 0;
+};
+
+export const folderName = (name: string) => {
+    return name.replace(/[/ ]/g, "-").toLowerCase();
+};
+
+export const fileName = (name: string) => {
```
Comment (BidishaMS) [inline src/client/powerpages/utils/CommonUtils.ts:14]: Can you write a comment which shows how the input/output looks like

File: src/client/powerpages/Contentsnippet.ts (hunk 2 above)
Comment (amitjoshi438) [reply src/client/powerpages/Contentsnippet.ts:21]: I thought proper nouns are not localized, do we localize them also?

File: src/client/powerpages/Contentsnippet.ts (hunk 4 above)
Comment (amitjoshi438) [reply src/client/powerpages/Contentsnippet.ts:65]: some properties are generic but as such it is specific to entities

File: src/client/powerpages/utils/MultiStepInput.ts (hunk 5 above)
Comment (amitjoshi438) [reply src/client/powerpages/utils/MultiStepInput.ts:126]: I have updated some custom method names but for vscode related methods I tried sticking to their docs and example implementations. Let me know if some specific method name should be changed.

File: src/client/powerpages/utils/CommonUtils.ts (hunk 7 above)
Comment (amitjoshi438) [reply src/client/powerpages/utils/CommonUtils.ts:6]: right, but here '!str' checks for all kind of **falsy** values : false, 0, "", null, undefined, NaN. Let me know if I am missing something.

**Pull request closed without merging**